File: src/main/java/no/fintlabs/mapping/InstanceMappingService.java
package no.fintlabs.mapping;

import no.fintlabs.model.configuration.CollectionMapping;
import no.fintlabs.model.configuration.FromCollectionMapping;
import no.fintlabs.model.configuration.ObjectMapping;
import no.fintlabs.model.configuration.ValueMapping;
import no.fintlabs.model.instance.InstanceObject;
import org.apache.commons.lang3.function.TriFunction;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Service
public class InstanceMappingService {

    private final InstanceReferenceService instanceReferenceService;
    private final ValueMappingService valueMappingService;


    public InstanceMappingService(InstanceReferenceService instanceReferenceService, ValueMappingService valueMappingService) {
        this.instanceReferenceService = instanceReferenceService;
        this.valueMappingService = valueMappingService;
    }

    public Map<String, ?> toMappedInstanceObject(
            ObjectMapping objectMapping,
            InstanceObject instance
    ) {
        return toMappedInstanceObject(objectMapping, instance, new InstanceObject[]{});
    }

    private Map<String, ?> toMappedInstanceObject(
            ObjectMapping objectMapping,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return Stream.of(
                        toMappedInstanceValuePerKey(
                                objectMapping.getValueMappingPerKey(),
                                instance,
                                selectedCollectionObjectsByCollectionIndex
                        ),
                        toMappedInstanceObjectPerKey(
                                objectMapping.getObjectMappingPerKey(),
                                instance,
                                selectedCollectionObjectsByCollectionIndex
                        ),
                        toMappedInstanceElementCollectionPerKey(
                                this::toMappedInstanceValue,
                                objectMapping.getValueCollectionMappingPerKey(),
                                instance,
                                selectedCollectionObjectsByCollectionIndex
                        ),
                        toMappedInstanceElementCollectionPerKey(
                                this::toMappedInstanceObject,
                                objectMapping.getObjectCollectionMappingPerKey(),
                                instance,
                                selectedCollectionObjectsByCollectionIndex
                        )
                )
                .map(Map::entrySet)
                .flatMap(Set::stream)
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
    }

    private Map<String, ?> toMappedInstanceValuePerKey(
            Map<String, ValueMapping> valueMappingPerKey,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return valueMappingPerKey.entrySet()
                .stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> toMappedInstanceValue(e.getValue(), instance, selectedCollectionObjectsByCollectionIndex)
                ));
    }

    private Object toMappedInstanceValue(
            ValueMapping valueMapping,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return valueMappingService.toValue(
                valueMapping,
                instance.getValuePerKey(),
                selectedCollectionObjectsByCollectionIndex
        );
    }

    private Map<String, Map<String, ?>> toMappedInstanceObjectPerKey(
            Map<String, ObjectMapping> objectMappingPerKey,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return objectMappingPerKey.entrySet()
                .stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> toMappedInstanceObject(e.getValue(), instance, selectedCollectionObjectsByCollectionIndex)
                ));
    }

    private <T, R> Map<String, Collection<R>> toMappedInstanceElementCollectionPerKey(
            TriFunction<T, InstanceObject, InstanceObject[], R> elementMappingFunction,
            Map<String, CollectionMapping<T>> collectionMappingPerKey,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        return collectionMappingPerKey.entrySet()
                .stream()
                .collect(Collectors.toMap(
                        Map.Entry::getKey,
                        e -> toMappedInstanceElements(elementMappingFunction, e.getValue(), instance, selectedCollectionObjectsByCollectionIndex)
                ));
    }

    private <T, R> Collection<R> toMappedInstanceElements(
            TriFunction<T, InstanceObject, InstanceObject[], R> elementMappingFunction,
            CollectionMapping<T> elementCollectionMapping,
            InstanceObject instance,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        Collection<T> elementMappings = elementCollectionMapping.getElementMappings();
        Collection<FromCollectionMapping<T>> fromCollectionMappings = elementCollectionMapping.getFromCollectionMappings();

        return Stream.concat(
                        elementMappings.stream()
                                .map(objectMapping -> elementMappingFunction.apply(objectMapping, instance, selectedCollectionObjectsByCollectionIndex)),
                        fromCollectionMappings.stream()
                                .map(fromCollectionMapping -> toMappedInstanceElements(
                                        elementMappingFunction,
                                        fromCollectionMapping.getElementMapping(),
                                        instance,
                                        fromCollectionMapping.getInstanceCollectionReferencesOrdered().toArray(new String[0]),
                                        0,
                                        selectedCollectionObjectsByCollectionIndex
                                ))
                                .flatMap(Collection::stream)
                )
                .collect(Collectors.toList());
    }

    private <T, R> Collection<R> toMappedInstanceElements(
            TriFunction<T, InstanceObject, InstanceObject[], R> elementMappingFunction,
            T elementMapping,
            InstanceObject instance,
            String[] instanceCollectionReferencesByCollectionIndex,
            int nextCollectionIndex,
            InstanceObject[] selectedCollectionObjectsByCollectionIndex
    ) {
        if (nextCollectionIndex == instanceCollectionReferencesByCollectionIndex.length) {
            return List.of(elementMappingFunction.apply(elementMapping, instance, selectedCollectionObjectsByCollectionIndex));
        }

        Collection<InstanceObject> nextCollection = instanceReferenceService.getInstanceObjectCollection(
                instanceCollectionReferencesByCollectionIndex[nextCollectionIndex],
                instance.getObjectCollectionPerKey(),
                selectedCollectionObjectsByCollectionIndex
        );

        return nextCollection
                .stream()
                .map(instanceObject -> {
                    InstanceObject[] newSelectedCollectionObjectsByCollectionIndex =
                            Arrays.copyOf(selectedCollectionObjectsByCollectionIndex, selectedCollectionObjectsByCollectionIndex.length + 1);
                    newSelectedCollectionObjectsByCollectionIndex[selectedCollectionObjectsByCollectionIndex.length] = instanceObject;
                    return toMappedInstanceElements(
                            elementMappingFunction,
                            elementMapping,
                            instance,
                            instanceCollectionReferencesByCollectionIndex,
                            nextCollectionIndex + 1,
                            newSelectedCollectionObjectsByCollectionIndex
                    );
                })
                .flatMap(Collection::stream)
                .toList();
    }
}

File: src/main/kotlin/no/novari/flyt/mapping/service/InstanceMappingService.kt
package no.novari.flyt.mapping.service

import no.novari.flyt.mapping.model.configuration.CollectionMapping
import no.novari.flyt.mapping.model.configuration.ObjectMapping
import no.novari.flyt.mapping.model.configuration.ValueMapping
import no.novari.flyt.mapping.model.instance.InstanceObject
import org.springframework.stereotype.Service

@Service
class InstanceMappingService(
    private val instanceReferenceService: InstanceReferenceService,
    private val valueMappingService: ValueMappingService,
) {
    fun toMappedInstanceObject(
        objectMapping: ObjectMapping,
        instance: InstanceObject,
    ): Map<String, Any?> {
        return toMappedInstanceObject(objectMapping, instance, emptyArray())
    }

    private fun toMappedInstanceObject(
        objectMapping: ObjectMapping,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Map<String, Any?> {
        return buildMap {
            putAll(
                toMappedInstanceValuePerKey(
                    objectMapping.valueMappingPerKey,
                    instance,
                    selectedCollectionObjectsByCollectionIndex,
                ),
            )
            putAll(
                toMappedInstanceObjectPerKey(
                    objectMapping.objectMappingPerKey,
                    instance,
                    selectedCollectionObjectsByCollectionIndex,
                ),
            )
            putAll(
                toMappedInstanceElementCollectionPerKey(
                    ::toMappedInstanceValue,
                    objectMapping.valueCollectionMappingPerKey,
                    instance,
                    selectedCollectionObjectsByCollectionIndex,
                ),
            )
            putAll(
                toMappedInstanceElementCollectionPerKey(
                    { nestedObjectMapping, instanceObject, selectedObjects ->
                        toMappedInstanceObject(nestedObjectMapping, instanceObject, selectedObjects)
                    },
                    objectMapping.objectCollectionMappingPerKey,
                    instance,
                    selectedCollectionObjectsByCollectionIndex,
                ),
            )
        }
    }

    private fun toMappedInstanceValuePerKey(
        valueMappingPerKey: Map<String, ValueMapping>,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Map<String, Any?> {
        return valueMappingPerKey.mapValues { (_, valueMapping) ->
            toMappedInstanceValue(valueMapping, instance, selectedCollectionObjectsByCollectionIndex)
        }
    }

    private fun toMappedInstanceValue(
        valueMapping: ValueMapping,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Any? {
        return valueMappingService.toValue(
            valueMapping,
            instance.valuePerKey,
            selectedCollectionObjectsByCollectionIndex,
        )
    }

    private fun toMappedInstanceObjectPerKey(
        objectMappingPerKey: Map<String, ObjectMapping>,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Map<String, Map<String, Any?>> {
        return objectMappingPerKey.mapValues { (_, objectMapping) ->
            toMappedInstanceObject(objectMapping, instance, selectedCollectionObjectsByCollectionIndex)
        }
    }

    private fun <T, R> toMappedInstanceElementCollectionPerKey(
        elementMappingFunction: (T, InstanceObject, Array<InstanceObject>) -> R,
        collectionMappingPerKey: Map<String, CollectionMapping<T>>,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Map<String, Collection<R>> {
        return collectionMappingPerKey.mapValues { (_, collectionMapping) ->
            toMappedInstanceElements(
                elementMappingFunction,
                collectionMapping,
                instance,
                selectedCollectionObjectsByCollectionIndex,
            )
        }
    }

    private fun <T, R> toMappedInstanceElements(
        elementMappingFunction: (T, InstanceObject, Array<InstanceObject>) -> R,
        elementCollectionMapping: CollectionMapping<T>,
        instance: InstanceObject,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Collection<R> {
        return buildList {
            addAll(
                elementCollectionMapping.elementMappings.map { elementMapping ->
                    elementMappingFunction(elementMapping, instance, selectedCollectionObjectsByCollectionIndex)
                },
            )

            elementCollectionMapping.fromCollectionMappings.forEach { fromCollectionMapping ->
                addAll(
                    toMappedInstanceElements(
                        elementMappingFunction,
                        fromCollectionMapping.elementMapping,
                        instance,
                        fromCollectionMapping.instanceCollectionReferencesOrdered.toTypedArray(),
                        0,
                        selectedCollectionObjectsByCollectionIndex,
                    ),
                )
            }
        }
    }

    private fun <T, R> toMappedInstanceElements(
        elementMappingFunction: (T, InstanceObject, Array<InstanceObject>) -> R,
        elementMapping: T,
        instance: InstanceObject,
        instanceCollectionReferencesByCollectionIndex: Array<String>,
        nextCollectionIndex: Int,
        selectedCollectionObjectsByCollectionIndex: Array<InstanceObject>,
    ): Collection<R> {
        if (nextCollectionIndex == instanceCollectionReferencesByCollectionIndex.size) {
            return listOf(
                elementMappingFunction(
                    elementMapping,
                    instance,
                    selectedCollectionObjectsByCollectionIndex,
                ),
            )
        }

        val nextCollection =
            instanceReferenceService.getInstanceObjectCollection(
                instanceCollectionReferencesByCollectionIndex[nextCollectionIndex],
                instance.objectCollectionPerKey,
                selectedCollectionObjectsByCollectionIndex,
            )

        return nextCollection.flatMap { instanceObject ->
            val newSelectedCollectionObjectsByCollectionIndex =
                selectedCollectionObjectsByCollectionIndex + instanceObject

            toMappedInstanceElements(
                elementMappingFunction,
                elementMapping,
                instance,
                instanceCollectionReferencesByCollectionIndex,
                nextCollectionIndex + 1,
                newSelectedCollectionObjectsByCollectionIndex,
            )
        }
    }
}
